### PR TITLE
Add Makefile documentation

### PR DESCRIPTION
**Pull request by ddelange**
#### Motivation

https://github.com/kserve/modelmesh-runtime-adapter/pull/38#discussion_r1193083493

#### Modifications

Add help texts

#### Result

```
all                       - Alias for `build`
build                     - Build runtime docker image
develop                   - Build develop docker image and run an interactive shell in the develop envionment
fmt                       - Run formatting
help                      - Print Makefile documentation
run                       - Build develop docker image and run a make command in the develop envionment (e.g. `make run fmt` will execute `make fmt` within the docker container)
test                      - Run tests
```

### CODE REVIEW

**Comment by ddelange** [comment]
cc @ckadner

**Comment by ckadner** [comment]
> cc @ckadner

Thanks @ddelange this is great. I will add a more detailed review soon.

**Comment by ddelange** [comment]
friendly ping @ckadner :)

File: Makefile [hunk 1]
```diff
@@ -21,35 +21,49 @@ ifeq (run,$(firstword $(MAKECMDGOALS)))
...
 build:
 	./scripts/build_docker.sh --target runtime
 
 .PHONY: build.develop
+## Build develop docker image
 build.develop:
 	./scripts/build_docker.sh --target develop
 
 .PHONY: develop
+## Build develop docker image and run an interactive shell in the develop envionment
```
Comment (ckadner) [inline Makefile:38]: may be shortened to:

```
## Run interactive shell inside developer container
```

File: Makefile [hunk 2]
```diff
@@ -21,35 +21,49 @@ ifeq (run,$(firstword $(MAKECMDGOALS)))
...
 develop: build.develop
 	./scripts/develop.sh
 
 .PHONY: run
+## Build develop docker image and run a make command in the develop envionment (e.g. `make run fmt` will execute `make fmt` within the docker container)
```
Comment (ckadner) [inline Makefile:43]: shorter:

```
## Run make target inside developer container (e.g. `make run fmt`)
```

File: Makefile [hunk 3]
```diff
@@ -21,35 +21,49 @@ ifeq (run,$(firstword $(MAKECMDGOALS)))
...
 run: build.develop
 	./scripts/develop.sh make $(RUN_ARGS)
 
 .PHONY: test
+## Run tests
 test:
 	./scripts/run_tests.sh
 
 .PHONY: fmt
+## Run formatting
```
Comment (ckadner) [inline Makefile:53]: might be good to tie this to the corresponding `lint` PR check?
```
## Auto-format source code and report code-style violations (lint)
```
or just
```
Format sources (lint)
```
?

File: Makefile [hunk 4]
```diff
@@ -21,35 +21,49 @@ ifeq (run,$(firstword $(MAKECMDGOALS)))
 endif
 
 .PHONY: all
+## Alias for `build`
 all: build
 
 .PHONY: build
+## Build runtime docker image
```
Comment (ckadner) [inline Makefile:28]: Docker uppercase

File: Makefile [hunk 5]
```diff
@@ -21,35 +21,49 @@ ifeq (run,$(firstword $(MAKECMDGOALS)))
...
 build:
 	./scripts/build_docker.sh --target runtime
 
 .PHONY: build.develop
+## Build develop docker image
```
Comment (ckadner) [inline Makefile:33]: hmm, maybe?

```
Build developer container image
```

File: Makefile [hunk 6]
```diff
@@ -21,35 +21,49 @@ ifeq (run,$(firstword $(MAKECMDGOALS)))
...
 .PHONY: proto.compile
+## Compile protos
 proto.compile:
 	./scripts/compile_protos.sh
 
+.PHONY: help
+## Print Makefile documentation
+help:
+	@perl -0 -nle 'printf("%-25s - %s\n", "$$2", "$$1") while m/^##\s*([^\r\n]+)\n^([\w-]+):[^=]/gm' $(MAKEFILE_LIST) | sort
+.DEFAULT_GOAL := help
```
Comment (ckadner) [inline Makefile:66]: maybe move this up to just above/below `.PHONY`

File: Makefile [hunk 7]
```diff
@@ -21,35 +21,49 @@ ifeq (run,$(firstword $(MAKECMDGOALS)))
...
 
 .PHONY: proto.compile
+## Compile protos
 proto.compile:
 	./scripts/compile_protos.sh
 
+.PHONY: help
+## Print Makefile documentation
+help:
+	@perl -0 -nle 'printf("%-25s - %s\n", "$$2", "$$1") while m/^##\s*([^\r\n]+)\n^([\w-]+):[^=]/gm' $(MAKEFILE_LIST) | sort
```
Comment (ckadner) [inline Makefile:65]: the regex needs to include `.` (dot) to print help for targets like `proto.compile` or `build.develop`

`'printf("%-25s - %s\n", "$$2", "$$1") while m/^##\s*([^\r\n]+)\n^([\w-]+):[^=]/gm'`    :x: 
`'printf("%-25s - %s\n", "$$2", "$$1") while m/^##\s*([^\r\n]+)\n^([\w.-]+):[^=]/gm'`  ✅ 
`                                                                      ^             `

**Review by ckadner — CHANGES_REQUESTED** [review]
Thank you @ddelange -- appologies for the long wait.

I have a few minor wording suggestions and a regex improvement :-)

**Comment by ckadner** [comment]
What do you think about shortening the first column and adding some color?

```Makefile
help:
	@perl -0 -nle 'printf("\033[36m  %-15s\033[0m %s\n", "$$2", "$$1") while m/^##\s*([^\r\n]+)\n^([\w.-]+):[^=]/gm' $(MAKEFILE_LIST) | sort

```

image

image

File: Makefile (hunk 1 above)
Comment (ddelange) [reply Makefile:38]:
```suggestion
## Run interactive shell inside developer container
```

File: Makefile (hunk 2 above)
Comment (ddelange) [reply Makefile:43]:
```suggestion
## Run make target inside developer container (e.g. `make run fmt`)
```

File: Makefile (hunk 3 above)
Comment (ddelange) [reply Makefile:53]:
```suggestion
## Auto-format source code and report code-style violations (lint)
```

File: Makefile (hunk 4 above)
Comment (ddelange) [reply Makefile:28]:
```suggestion
## Build runtime Docker image
```

File: Makefile (hunk 5 above)
Comment (ddelange) [reply Makefile:33]:
```suggestion
## Build developer container image
```

File: Makefile (hunk 7 above)
Comment (ddelange) [reply Makefile:65]:
```suggestion
	@perl -0 -nle 'printf("\033[36m  %-15s\033[0m %s\n", "$$2", "$$1") while m/^##\s*([^\r\n]+)\n^([\w.-]+):[^=]/gm' $(MAKEFILE_LIST) | sort
```

File: Makefile (hunk 6 above)
Comment (ddelange) [reply Makefile:66]:
```suggestion
```

Comment (ddelange) [inline Makefile:66]:
```suggestion
.DEFAULT_GOAL := help
.PHONY: help
## Print Makefile documentation
help:
	@perl -0 -nle 'printf("%-25s - %s\n", "$$2", "$$1") while m/^##\s*([^\r\n]+)\n^([\w-]+):[^=]/gm' $(MAKEFILE_LIST) | sort
```

**Comment by ddelange** [comment]
@ckadner done in all three PRs 👍

**Review by ckadner — APPROVED** [review]
Thank you @ddelange 

/lgtm

**Pull request merged**